qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Curve fitting functions for experiment analysis
+"""
+# pylint: disable = invalid-name
+
+from typing import List, Dict, Tuple, Callable, Optional
+
+import numpy as np
+import scipy.optimize as opt
+
+from qiskit.exceptions import QiskitError
+from qiskit_experiments.base_analysis import AnalysisResult
+from qiskit_experiments.analysis.data_processing import filter_data
+
+
+def curve_fit(
+    func: Callable,
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray] = None,
+    **kwargs,
+) -> AnalysisResult:
+    """Use non-linear least squares to fit a function, f, to data
+
+    Wraps scipy.optimize.curve_fit
+
+    Args:
+        func: a fit function `f(x *params)`.
+        xdata: a 1D array of x-data
+        ydata: a 1D array of y-data
```

Comment:
```suggestion
        xdata: a 1D float array of x-data
        ydata: a 1D float array of y-data
```
or supporting complex value fit by default would be great (just concatenating real and imag value regardless of float or complex).